README.md
# nokyc bot
A telegram bot to list all current [Bisq](https://bisq.network), [HodlHodl](https://hodlhodl.com) and [Robosats](https://unsafe.robosats.com) offers

# Instructions
- Change `config.py` with your bot token, tor proxy port, and webhook info (if you're going to use webhooks to connect to your bot)
- For development and testing, set your environment variable `MODE` to `'polling'`. For production, set the variable to `'webhook'`
- Run the bot with `python3 bot.py`
  



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
  
## v0.1.0 (26/07/2022)

## Added
- Multilang interface (en, es and zh are now supported)
- Improved interface in conversational style
- More currencies

## v0.0.1 (07/04/2022)

## Added
- config.py template
- error log for each exchange API
  
## Fixed
- Prices are now obtained from Bisq instead of coindesk
  

## v0.0.1 (07/04/2022)

- Initial release
  


utils/utils.py
import logging
import requests
import io

# Exchange APIs
from exchanges.bisq import Bisq
from exchanges.robosats import Robosats
from exchanges.hodlhodl import HodlHodl

import config as config

import prettytable as pt
from PIL import Image, ImageDraw, ImageFont

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)


def get_tor_session():
    logging.info("starting tor session")
    session = requests.session()
    session.proxies = {'http':  'socks5h://127.0.0.1:' + config.TOR_PORT,
                       'https': 'socks5h://127.0.0.1:' + config.TOR_PORT}
    return session


def print_orders(fiat, direction, limit, exchanges):
    """Get orders from bisq, hodlhodl and robosats according to parameters

    Args:
        fiat (string): usd, eur, ...
        direction (string): 'buy' or 'sell'
        limit (int): percentage of premium
        exchanges (list): exchanges to query
    """
    logging.info('Exchanges: ' + exchanges)
    session = get_tor_session()
    price_exch = Bisq.getFiatPrice(fiat, session)
    if exchanges == "all":
        logging.info("Obtaining orders from bisq...")
        bisqOffers = Bisq.getOffers(fiat, direction, price_exch, session)
        logging.info("Obtaining orders from robosats...")
        robosatsOffers = Robosats.getOffers(fiat, direction, session)
        logging.info("Obtaining orders from hodlhodl...")
        hodlhodlOffers = HodlHodl.getOffers(
            fiat, direction, price_exch, session)
        allOffers = bisqOffers + robosatsOffers + hodlhodlOffers
    elif exchanges == "bisq":
        logging.info("Obtaining orders from bisq...")
        bisqOffers = Bisq.getOffers(fiat, direction, price_exch, session)
        allOffers = bisqOffers
    elif exchanges == "hodlhodl":
        logging.info("Obtaining orders from hodlhodl...")
        hodlhodlOffers = HodlHodl.getOffers(
            fiat, direction, price_exch, session)
        allOffers = hodlhodlOffers
    elif exchanges == "robosats":
        logging.info("Obtaining orders from robosats...")
        robosatsOffers = Robosats.getOffers(fiat, direction, session)
        allOffers = robosatsOffers
    if direction == "buy":
        allOffers.sort(key=lambda item: item.get('price'), reverse=True)
    elif direction == "sell":
        allOffers.sort(key=lambda item: item.get('price'))

    table = pt.PrettyTable(
        ['Exchange', 'Price', 'Dif', 'Min', 'Max', 'Method'])

    for offer in allOffers:
        if offer['method'].lower() not in config.avoid_methods:
            row = [f"{offer['exchange']:10}", f"{offer['price']:8n}", f"{offer['dif']:4.1f}%",
                   f"{offer['min_amount']:7n}", f"{offer['max_amount']:7n}", f"{offer['method']}"]
            if limit == "alloffers":
                table.add_row(row)
            else:
                if (direction == "buy") and (offer['dif'] > int(limit)):
                    table.add_row(row)
                if (direction == "sell") and (offer['dif'] < int(limit)):
                    table.add_row(row)
        # TODO: split the message in chunks so it won't exceed the max 4096 characters / msg
        if len(table.get_string()) > 3800:
            logger.info("limite de caracteres alcanzado")
            break
    logging.info("Done!")
    return(price_exch, table)


def table_to_img(table):
    """Convert table to image

    Args:
        table (PrettyTable): Table to convert

    Returns:
        Image: image
    """
    # Calculate dimensions of image from text
    fnt = ImageFont.truetype("fonts/FreeMono.ttf", 15)
    img = Image.new('RGB', (200, 100))
    d = ImageDraw.Draw(img)
    d.text((10, 10), table, font=fnt, fill=(255, 0, 0))
    text_width, text_height = d.textsize(table, font=fnt)
    # draw the actal image
    img = Image.new('RGB', (text_width+20, text_height+20), (255, 255, 255))
    d = ImageDraw.Draw(img)
    d.text(xy=(10, 10), text=table, font=fnt, fill=(0, 0, 0))
    s = io.BytesIO()
    img.save(s, 'png')
    s.seek(0)
    return s


exchanges/bisq.py
#!/usr/bin/env python3

import json
import logging
import requests

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)

class Bisq:
    
    def getOffers(fiat, direction, refprice, session):

        # fiat = eur, usd, 
        # direction = buy or sell
        # refprice = int
        # tor = 1 or 0

        bisqBaseUrlTor = 'http://bisqmktse2cabavbr2xjq7xw3h6g5ottemo5rolfcwt6aly6tp5fdryd.onion'

        bisqApi = f"{bisqBaseUrlTor}/api/offers?market=btc_{fiat.upper()}&direction={direction.upper()}"
        try:
            f = session.get(bisqApi)
        except requests.exceptions.Timeout as e:
            # Maybe set up for a retry, or continue in a retry loop
            logger.error("Error obtaining orders from Bisq (timeout): %s - %s" % (e.errno, e.strerror))
            return []
        except requests.exceptions.TooManyRedirects as e:
            logger.error("Error obtaining orders from Bisq (too many redirects): %s - %s" % (e.errno, e.strerror))
            return []
        except requests.exceptions.RequestException as e:
            logger.error("Error obtaining orders from Bisq: %s - %s" % (e.errno, e.strerror))
            return []
        try:
            values = f.json()
            f.close()
        except json.decoder.JSONDecodeError as e:
            logger.error("Error decoding orders from Bisq: %s - %s" % (e.errno, e.strerror))
            return []
        key = f"btc_{fiat}"
        
        alloffers = []

        for line in values[key][direction + 's' ]:
            offer = {}
            offer['exchange'] = 'Bisq'
            offer['price'] = int(float(line['price']))
            offer['dif'] = (offer['price']/refprice-1)*100
            offer['min_btc'] = float(line['min_amount'])
            offer['max_btc'] = float(line['amount'])
            offer['min_amount'] = int(offer['min_btc'] * offer['price'])
            offer['max_amount'] = int(float(line['volume']))
            offer['method'] = line['payment_method']
            alloffers.append(offer)
        alloffers.sort(key=lambda item: item.get('price'))
        return alloffers
    
    def getFiatPrice(fiat, session):
        bisqApi = 'http://wizpriceje6q5tdrxkyiazsgu7irquiqjy2dptezqhrtu7l2qelqktid.onion/getAllMarketPrices'
        try:
            f = session.get(bisqApi)
        except requests.exceptions.Timeout as e:
            # Maybe set up for a retry, or continue in a retry loop
            logger.error("Error obtaining price from Bisq (timeout): %s - %s" % (e.errno, e.strerror))
            return 1
        except requests.exceptions.TooManyRedirects as e:
            logger.error("Error obtaining price from Bisq (too many redirects): %s - %s" % (e.errno, e.strerror))
            return 1
        except requests.exceptions.RequestException as e:
            logger.error("Error obtaining price from Bisq: %s - %s" % (e.errno, e.strerror))
            return 1 
            # * I return 1 instead of 0 to avoid errors raised dividing by the price    
        
        try:    
            priceapi = f.json()
            f.close()
        except json.decoder.JSONDecodeError as e:
            logger.error("Error decoding orders from Bisq: %s - %s" % (e.errno, e.strerror))
            return 1

        for currency in priceapi['data']:
            if (currency['currencyCode'].lower()==fiat):
                return int(float(currency['price']))


exchanges/fiat.py
#!/usr/bin/env python3

import json
import requests

def getcgprice(fiat, session):
    coingeckoApi = 'https://api.coingecko.com/api/v3/coins/markets?vs_currency='
    
    f = session.get(coingeckoApi + fiat + '&ids=bitcoin')
    priceapi = f.json()
    f.close()
    price = priceapi[0]['current_price']
    return price

def getkprice(fiat, session):
    # krakenApi = 'https://api.kraken.com/0/public/Ticker?pair=XBT'
    # coindeskApi = 'http://api.coindesk.com/v1/bpi/currentprice.json'
    try:
        response = requests.get(
            'http://api.coindesk.com/v1/bpi/currentprice.json')
        data = response.json()
        price = int(float(data["bpi"][fiat.upper()]["rate"].replace(",", "")))
    except requests.exceptions.RequestException:
        print("Err obtaining price")
        price = 0
    
    # f = session.get(krakenApi + fiat.upper())
    # priceapi = f.json()
    # f.close()
    
    # if fiat in ['eur', 'usd', 'gbp', 'cad']:
    #     key = 'XXBTZ' + fiat.upper()
    # else:
    #     key = 'XBT' + fiat.upper()
    # price = int(float(priceapi['result'][key]['c'][0]))
    return price

class Fiat:
    def getfiatprice(fiat, session):
        fiat_kraken = ['eur', 'usd', 'gbp', 'cad', 'aud', 'chf']
        fiat_coingecko = ['brl', 'czk', 'sek', 'nzd', 'dkk', 'pln']
    
        if fiat in fiat_kraken:
            price = getkprice(fiat, session)
        else:
            price = getcgprice(fiat, session)

        return price



exchanges/hodlhodl.py
#!/usr/bin/env python3

import json
import requests
import logging

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)


class HodlHodl:

    def getOffers(curr, direction, refprice, session):
        curr = curr.upper()
        api = f"https://hodlhodl.com/api/v1/offers?filters[side]={direction}&filters[include_global]=true&filters[currency_code]={curr}&filters[only_working_now]=true&sort[by]=price"
        try:
            f = session.get(api)
        except requests.exceptions.Timeout as e:
        # Maybe set up for a retry, or continue in a retry loop
            logger.error("Error obtaining orders from Robosats (timeout): %s - %s" % (e.errno, e.strerror))
            return []
        except requests.exceptions.TooManyRedirects as e:
            logger.error("Error obtaining orders from Robosats (too many redirects): %s - %s" % (e.errno, e.strerror))
            return []
        except requests.exceptions.RequestException as e:
            logger.error("Error obtaining orders from Robosats: %s - %s" % (e.errno, e.strerror))
            return []

        try:
            jsonweb = f.json()
            f.close()
            alloffers = jsonweb['offers']
        except json.decoder.JSONDecodeError as e:
            logger.error("Error decoding orders from HodlHodl: %s - %s" % (e.errno, e.strerror))
            return []

        lista = []

        for offer in alloffers:
            offers = {}
            offers['exchange'] = "HodlHodl"
            offers['price'] = int(float(offer['price']))
            offers['dif'] = (offers['price']/refprice - 1)*100
            offers['currency'] = offer['currency_code']
            offers['min_amount'] = int(float(offer['min_amount']))
            offers['max_amount'] = int(float(offer['max_amount']))
            offers['min_btc'] = offers['min_amount']/offers['price']
            offers['max_btc'] = offers['max_amount']/offers['price']
            status = offer['trader']['online_status']
            if (direction == "buy"):
                offers['method'] = offer['payment_methods'][0]['name']
            else:
                offers['method'] = offer['payment_method_instructions'][0]['payment_method_name']
            if "SEPA" in offers['method']:
                offers['method'] = "SEPA"
            elif "Any national bank" in offers['method']:
                offers['method'] = "NATIONAL_BANK"
            if (status == 'online'):
                lista.append(offers)

        lista.sort(key=lambda item: item.get("price"))
        return lista


exchanges/robosats.py
#!/usr/bin/env python3

import json
import requests
import logging

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)

currencies = {"1":"USD", "2":"EUR", "3":"JPY", "4":"GBP", "5":"AUD", "6":"CAD", "7":"CHF", "8":"CNY", "9":"HKD", "10":"NZD", "11":"SEK", "12":"KRW", "13":"SGD", "14":"NOK",
	"15":"MXN", "16":"KRW", "17":"RUB", "18":"ZAR", "19":"TRY", "20":"BRL", "21":"CLP", "22":"CZK", "23":"DKK", "24":"HRK", "25":"HUF", "26":"INR", "27":"ISK", "28":"PLN",
	"29":"RON", "30":"ARS", "31":"VES", "32":"COP", "33":"PEN", "34":"UYU", "35":"PYG", "36":"BOB", "37":"IDR", "38":"ANG", "39":"CRC", "40":"CUP", "41":"DOP", "42":"GHS",
	"43":"GTQ", "44":"ILS", "45":"JMD", "46":"KES", "47":"KZT", "48":"MYR", "49":"NAD", "50":"NGN", "51":"AZN", "52":"PAB", "53":"PHP", "54":"PKR", "55":"QAR", "56":"SAR", 
        "57":"THB", "58":"TTD", "59":"VND", "60":"XOF", "61":"TWD", "300":"XAU", "1000":"BTC"}

class Robosats:
    def getOffers(fiat, direction, session):

        robosatsTor = 'http://robosats6tkf3eva7x2voqso3a5wcorsnw34jveyxfqi2fu7oyheasid.onion'
 
        key_list = list(currencies.keys())
        val_list = list(currencies.values())
        position = val_list.index(fiat.upper())
        currency = key_list[position]

        if (direction=="sell"):
            typeoffer = 1
        else:
            typeoffer = 0

        command = f'/api/book/?currency={currency}&type={typeoffer}'

        try:
            f = session.get(robosatsTor + command)
        except requests.exceptions.Timeout as e:
            # Maybe set up for a retry, or continue in a retry loop
            logger.error("Error obtaining orders from Robosats (timeout): %s - %s" % (e.errno, e.strerror))
            return []
        except requests.exceptions.TooManyRedirects as e:
            logger.error("Error obtaining orders from Robosats (too many redirects): %s - %s" % (e.errno, e.strerror))
            return []
        except requests.exceptions.RequestException as e:
            logger.error("Error obtaining orders from Robosats: %s - %s" % (e.errno, e.strerror))
            return []
        
        try:
            values = f.json()
            f.close()
        except json.decoder.JSONDecodeError as e:
            logger.error("Error decoding orders from Robosats: %s - %s" % (e.errno, e.strerror))
            return []
        alloffers = []

        for line in values:
            if (line == "not_found"):
                break
            offer = {}
            offer['exchange'] = 'Robosats'
            offer['price'] = int(float(line['price']))
            offer['dif'] = float(line['premium'])
            if (line['amount'] is not None):
                offer['min_amount'] = int(float(line['amount']))
                offer['max_amount'] = int(float(line['amount']))
            else:
                offer['min_amount'] = int(float(line['min_amount']))
                offer['max_amount'] = int(float(line['max_amount']))
            offer['min_btc']=offer['min_amount']/offer['price']
            offer['max_btc']=offer['max_amount']/offer['price']
            offer['method']=line['payment_method']
            alloffers.append(offer)
        alloffers.sort(key=lambda item: item.get('price'))

        return alloffers


bot.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""This is a telegram bot to get offers from p2p exchanges like
    bisq, hodlhodl and robosats"""

import logging

from telegram.ext.updater import Updater
from telegram.update import Update
from telegram.ext.callbackcontext import CallbackContext
from telegram.ext.commandhandler import CommandHandler
from telegram.ext.messagehandler import MessageHandler
from telegram.ext.filters import Filters
from telegram.ext import CallbackQueryHandler
from telegram import InlineKeyboardButton, InlineKeyboardMarkup, ParseMode

from exchanges.bisq import Bisq
from exchanges.robosats import Robosats
from exchanges.hodlhodl import HodlHodl

import requests
import os
import prettytable as pt
from PIL import Image, ImageDraw, ImageFont
import io

import config as config

# read MODE env variable, fall back to 'polling' when undefined
mode = os.environ.get("MODE", config.DEFAULT_CONNECTION)

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)

# Keyboards
keyboard_actions = [
    [
        InlineKeyboardButton("I want to buy BTC", callback_data='sell'),
        InlineKeyboardButton("I want to sell BTC", callback_data='buy')
    ]
]

keyboard_exchanges = [
    [
        InlineKeyboardButton("Bisq", callback_data='bisq'),
        InlineKeyboardButton("HodlHodl", callback_data='hodlhodl'),
        InlineKeyboardButton("Robosats", callback_data='robosats'),
        InlineKeyboardButton("All", callback_data='all')
    ]
]

keyboard_currencies = [
    [
        InlineKeyboardButton("USD", callback_data='usd'),
        InlineKeyboardButton("EUR", callback_data='eur')
    ]
]

keyboard_premium = [
    [
        InlineKeyboardButton("-9%", callback_data='-9'),
        InlineKeyboardButton("-7%", callback_data='-7'),
        InlineKeyboardButton("-5%", callback_data='-5'),
        InlineKeyboardButton("-3%", callback_data='-3'),
        InlineKeyboardButton("-1%", callback_data='-1')
    ],
    [
        InlineKeyboardButton("Show all", callback_data='alloffers'),
    ],
    [
        InlineKeyboardButton("0%", callback_data='0'),
        InlineKeyboardButton("1%", callback_data='1'),
        InlineKeyboardButton("3%", callback_data='3'),
        InlineKeyboardButton("5%", callback_data='5'),
        InlineKeyboardButton("7%", callback_data='7'),
        InlineKeyboardButton("9%", callback_data='9'),
    ]
]
keyboard_developer = [
    [InlineKeyboardButton("Message the developer",
                          url='https://t.me/fgbernal')]
]

keyboard_runquery = [
    [InlineKeyboardButton("Run query", callback_data='query')]
]

keyboard_format = [
    [
        InlineKeyboardButton("Plain text", callback_data='text'),
        InlineKeyboardButton("Image", callback_data='img')
    ]
]

# End User configuration


def table_to_img(table):
    # Calculate dimensions of image from text
    fnt = ImageFont.truetype("fonts/FreeMono.ttf", 15)
    img = Image.new('RGB', (200, 100))
    d = ImageDraw.Draw(img)
    d.text((10, 10), table, font=fnt, fill=(255, 0, 0))
    text_width, text_height = d.textsize(table, font=fnt)
    # draw the actal image
    img = Image.new('RGB', (text_width+20, text_height+20), (255, 255, 255))
    d = ImageDraw.Draw(img)
    d.text(xy=(10, 10), text=table, font=fnt, fill=(0, 0, 0))
    s = io.BytesIO()
    img.save(s, 'png')
    s.seek(0)
    return s


def get_tor_session():
    logging.info("starting tor session")
    session = requests.session()
    session.proxies = {'http':  'socks5h://127.0.0.1:' + config.TOR_PORT,
                       'https': 'socks5h://127.0.0.1:' + config.TOR_PORT}
    return session


def start(update: Update, context: CallbackContext):
    update.message.reply_text(
        'Welcome to noKYCbot. I will help you find buy and sell orders in Bisq, HodlHodl and Robosats. Please configure the following options:')
    action_url(update, context)
    exchange_url(update, context)
    currency_url(update, context)
    premium_url(update, context)
    format_url(update, context)
    # TODO: Add button to run query from this menu


def help(update: Update, context: CallbackContext):
    reply_markup = InlineKeyboardMarkup(keyboard_developer)
    update.message.reply_text("""Available Commands:
    /start - configure your options
    /query - execute the query
    /help - show this help message
    /exchange - select the exchange
    /currency - set the currency you want to pay or get paid
    /premium - set the premium over or under market price you're willing to accept
    /action - you want to buy or sell bitcoin?
    /format - output format, image or plain text
    """, reply_markup=reply_markup)


def print_orders(fiat, direction, limit, exchanges):
    session = get_tor_session()
    price_exch = Bisq.getFiatPrice(fiat, session)
    if exchanges == "all":
        logging.info("Obtaining orders from bisq")
        bisqOffers = Bisq.getOffers(fiat, direction, price_exch, session)
        logging.info("Obtaining orders from robosats")
        robosatsOffers = Robosats.getOffers(fiat, direction, session)
        logging.info("Obtaining orders from hodlhodl")
        hodlhodlOffers = HodlHodl.getOffers(
            fiat, direction, price_exch, session)
        allOffers = bisqOffers + robosatsOffers + hodlhodlOffers
    elif exchanges == "bisq":
        logging.info("Obtaining orders from bisq")
        bisqOffers = Bisq.getOffers(fiat, direction, price_exch, session)
        allOffers = bisqOffers
    elif exchanges == "hodlhodl":
        logging.info("Obtaining orders from hodlhodl")
        hodlhodlOffers = HodlHodl.getOffers(
            fiat, direction, price_exch, session)
        allOffers = hodlhodlOffers
    elif exchanges == "robosats":
        logging.info("Obtaining orders from robosats")
        robosatsOffers = Robosats.getOffers(fiat, direction, session)
        allOffers = robosatsOffers
    if direction == "buy":
        allOffers.sort(key=lambda item: item.get('price'), reverse=True)
    elif direction == "sell":
        allOffers.sort(key=lambda item: item.get('price'))

    table = pt.PrettyTable(
        ['Exchange', 'Price', 'Dif', 'Min', 'Max', 'Method'])

    for offer in allOffers:
        if offer['method'].lower() not in config.avoid_methods:
            row = [f"{offer['exchange']:10}", f"{offer['price']:8n}", f"{offer['dif']:4.1f}%",
                   f"{offer['min_amount']:7n}", f"{offer['max_amount']:7n}", f"{offer['method']}"]
            if limit == "alloffers":
                table.add_row(row)
            else:
                if (direction == "buy") and (offer['dif'] > int(limit)):
                    table.add_row(row)
                if (direction == "sell") and (offer['dif'] < int(limit)):
                    table.add_row(row)
        # TODO: split the message in chunks so it won't exceed the max 4096 characters / msg
        if len(table.get_string()) > 3800:
            logger.info("limite de caracteres alcanzado")
            break
    return(price_exch, table)


def query_url(update: Update, context: CallbackContext):
    if "exchange" in context.user_data.keys():
        exchange = context.user_data["exchange"]
    else:
        exchange = "all"

    if "action" in context.user_data.keys():
        action = context.user_data["action"]
    else:
        action = "buy"

    if "premium" in context.user_data.keys():
        premium = context.user_data["premium"]
    else:
        premium = "alloffers"

    if "currency" in context.user_data.keys():
        fiat = context.user_data["currency"]
    else:
        fiat = "eur"

    if "format" in context.user_data.keys():
        format = context.user_data["format"]
    else:
        format = 'text'

    price, result = print_orders(fiat, action, premium, exchange)
    msg = f"BTC price: {price} {fiat.upper()}\nBTC {action} offers:\n" + \
        f"{result}"

    if format == 'img':
        img = table_to_img(msg)
        update.message.reply_photo(img)
    else:
        update.message.reply_text(
            f'```{msg}```', parse_mode=ParseMode.MARKDOWN_V2)


def button(update: Update, context: CallbackContext):
    update.callback_query.answer()
    if update.callback_query.data in ['bisq', 'hodlhodl', 'robosats', 'all']:
        context.user_data["exchange"] = update.callback_query.data
        update.callback_query.bot.send_message(
            chat_id=update.callback_query.from_user.id,
            text="Ok, I will show you orders from " + context.user_data["exchange"] + " exchange(s)")
    elif update.callback_query.data in ['usd', 'eur']:
        context.user_data["currency"] = update.callback_query.data
        update.callback_query.bot.send_message(
            chat_id=update.callback_query.from_user.id,
            text="Currency is set to " + context.user_data["currency"])
    elif update.callback_query.data in ['buy', 'sell']:
        context.user_data["action"] = update.callback_query.data
        update.callback_query.bot.send_message(
            chat_id=update.callback_query.from_user.id,
            text="Got it!, I will show you " + context.user_data["action"] + " offers from other users")
    elif update.callback_query.data in ['1', '3', '5', '7', '9', '-1', '-3', '-5', '-7', '-9', 'alloffers']:
        context.user_data["premium"] = update.callback_query.data
        if context.user_data["premium"] == "alloffers":
            msg = "All right, I will show you all orders, regardless of price"
        elif int(context.user_data["premium"]) < 0:
            msg = "ok, I will set the price limit at " + \
                context.user_data["premium"] + "% below the market price."
        elif int(context.user_data["premium"]) > 0:
            msg = "ok, I will set the price limit at " + \
                context.user_data["premium"] + "% above the market price."
        update.callback_query.bot.send_message(
            chat_id=update.callback_query.from_user.id,
            text=msg)
    elif update.callback_query.data in ['img', 'text']:
        context.user_data["format"] = update.callback_query.data
        update.callback_query.bot.send_message(
            chat_id=update.callback_query.from_user.id,
            text="Ok, I will display the result as " + context.user_data["format"])
    elif update.callback_query.data in ['query']:
        update.callback_query.bot.send_message(
            chat_id=update.callback_query.from_user.id,
            text="Executing query. Please wait...")
        query_url(update, context)


def exchange_url(update: Update, context: CallbackContext):
    reply_markup = InlineKeyboardMarkup(keyboard_exchanges)
    update.message.reply_text(
        'Which exchange(s) do you want to use?', reply_markup=reply_markup)


def currency_url(update: Update, context: CallbackContext):
    reply_markup = InlineKeyboardMarkup(keyboard_currencies)
    update.message.reply_text(
        'In which currency do you want to pay or get paid?', reply_markup=reply_markup)


def premium_url(update: Update, context: CallbackContext):
    reply_markup = InlineKeyboardMarkup(keyboard_premium)
    update.message.reply_text(
        'Which premium are you willing to accept over or under market price?', reply_markup=reply_markup)


def action_url(update: Update, context: CallbackContext):
    reply_markup = InlineKeyboardMarkup(keyboard_actions)
    update.message.reply_text(
        'What do you want to do?', reply_markup=reply_markup)


def format_url(update: Update, context: CallbackContext):
    reply_markup = InlineKeyboardMarkup(keyboard_format)
    update.message.reply_text(
        'In which format do you want the result to be displayed?', reply_markup=reply_markup)


def unknown_text(update: Update, context: CallbackContext):
    update.message.reply_text(
        "Sorry I can't recognize what you said '%s'" % update.message.text)


def main() -> None:

    updater = Updater(config.TOKEN,
                      use_context=True)
    disp = updater.dispatcher
    disp.add_handler(CommandHandler('start', start))
    disp.add_handler(CommandHandler('help', help))
    disp.add_handler(CommandHandler('exchange', exchange_url))
    disp.add_handler(CommandHandler('currency', currency_url))
    disp.add_handler(CommandHandler('premium', premium_url))
    disp.add_handler(CommandHandler('action', action_url))
    disp.add_handler(CommandHandler('format', format_url))
    disp.add_handler(CommandHandler('query', query_url))

    disp.add_handler(CallbackQueryHandler(button))

    # Filters out unknown messages.
    disp.add_handler(MessageHandler(Filters.text, unknown_text))

    if mode == 'webhook':
        PORT = os.environ.get("PORT", config.DEFAULT_CONNECTION)
        logger.info("starting webhook")
        updater.start_webhook(listen="0.0.0.0", port=PORT, url_path=config.TOKEN,
                              webhook_url=config.APP_NAME + config.TOKEN)
    else:
        updater.start_polling()
        updater.idle()


if __name__ == '__main__':
    main()
